enhancements/windows-containers/windows-machine-config-operator.md

```diff
@@ -0,0 +1,177 @@
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Provide workflows for adding Windows instances provisioned using the Machine
+  API to OpenShift clusters
+* Perform all the required steps within the VM for it to be added to the cluster
+  as an OpenShift worker node
+
+### Non-Goals
+
+As part of this enhancement we do not plan to support:
+* Installing the container runtime in the Windows node
+* Upgrades (This will be addressed in a separate enhancement)
+  * Windows operating system upgrades
+  * Node management like k8s compute component upgrades, reboots and node
+    draining
+* OpenShift Builds
+* OpenShift DeploymentConfigs
+
+## Proposal
+
+The two main pieces that will allow us to achieve our goals is the Windows
+Machine Config operator (WMCO) and a Windows Machine Config Bootstrapper (WMCB)
+binary. The operator will collect information from the cluster and transfer it
+to the Windows node. The WMCB will then perform the necessary steps on the node
+for it to join the cluster. [WMCB](https://github.com/openshift/windows-machine-config-bootstrapper)
+has already been developed as part of the Windows Containers
+[dev preview](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/tools/ansible/docs/ocp-4-4-with-windows-server.md).
+
+### Justification
+
+The reason we are using a WMCB binary rather than a container image is
+that all Windows container images are packaged with a Windows kernel and Red Hat
+has a policy to not ship 3rd party kernels for support reasons.
+
+### Design Details
+
+The operator image will be packaged with the following binaries:
+* Kubelet
+* CNI plugins
+* Hybrid-overlay
+* WMCB
+* Kube-proxy
+
+WMCO will be published to OperatorHub as a Red Hat operator. The way the cluster
+admin will enable Windows workloads on their cluster would be to first install
+WMCO. The operator expects the cluster admin to create a predertmined Secret in
+its namespace containing a private key that will be used to interact with the
+Windows instance. The operator will check for this Secret during boot up time
+and create a user data Secret which the cluster admin can reference in the
+Windows MachineSet that they create as described below. The operator will
+populate the user data secret with a public key that corresponds to the private
+key. In addition it will have all the pre-requisite steps needed for setting the
+VM up for key based SSH connections.
+
+To express that they want to bring up Windows worker nodes, they will
+create a MachineSet where they reference a Windows OS image which has the
+Docker container runtime add-on enabled. They will also specify the user data
+secret created by the operator as described above. The operator will watch for
+Machine objects with a special label that indicate that they map to Windows VMs.
+Once the Machine object is in the "provisioned" state, the operator will connect
```

Comment:
Connect is hard to comprehend perhaps `once the VM associated with the machine object is accessible from WMCO, WMCO transfers the binaries...`?

Reply:
I am not sure why connect is so hard to comprehend but I will add the associated VM bit.